File: game_dutch.py
from game import Game
from template import Template


class Dutch(Game):
    def __init__(self):
        super().__init__()
        self.context = 'Nederlands'
        self.multiplier = 10
        self.menu_template = Template(
            'Menu\n',
            '----\n',
            '[1] Contract tonen\n',
            '[2] Specificatie tonen\n',
            '[3] Unit test toevoegen\n',
            '[4] Hint voor unit test tonen (-{penalty_hint})\n',
            '[5] Unit testen inleveren (-{penalty_bug} bij foutmelding)\n',
            '[0] Einde (-{penalty_end} bij foutmelding)\n',
        )
        self.choice_template = Template(
            'Keuze',
        )
        self.contract_template = Template(
            'Contract\n',
            '--------\n',
            'Wij laten een nieuwe functie ontwikkelen door een extern softwarebedrijf.',
            'We hebben jou ingehuurd om te zorgen dat die functie ALTIJD het juiste resultaat geeft.',
            'Wat goede resultaten zijn staat beschreven in de specificatie.',
            'Jouw taak is om voldoende unit testen te schrijven voor die functie,',
            'zodat de functie geen foute resultaten meer kan geven.',
            'Voor het hele traject krijg je {initial_score}.',
            'In het menu staan bij sommige acties kosten vermeld voor jou.',
            'Als een gebruiker bijvoorbeeld een fout constateert in een functie die slaagt voor al jouw unit testen,',
            'dan betaal jij een boete van {penalty_bug}.',
        )
        self.score_template = Template(
            'Verdiensten\n',
            '-----------\n',
            '{score}',
        )
        self.no_unit_tests_template = Template(
            'Unit testen\n',
            '-----------\n',
            '(Je hebt nog geen unit test geschreven)',
        )
        self.unit_tests_template = Template(
            'Unit testen\n',
            '-----------\n',
            '{unit_tests}',
        )
        self.current_function_template = Template(
            'Huidige functie\n',
            '---------------\n',
            '{worst_passing_function}',
        )
        self.hint_unit_test_template = Template(
            'Hint voor unit test\n',
            '-------------------\n',
            'Een unit test die nu niet slaagt is bijvoorbeeld de volgende.',
            '{failing_unit_test}',
            'De kosten voor deze hint zijn {penalty_hint}.',
        )
        self.add_unit_test_template = Template(
            'Unit test toevoegen\n',
            '-------------------',
        )
        self.useless_unit_test_template = Template(
            'Unit test toegevoegd\n',
            '--------------------\n',
            'We hebben deze unit test toegevoegd.',
            'Je unit test lijkt erg veel op een eerdere unit test.',
            'We denken daarom dat deze unit test niet zo zinvol is.',
        )
        self.useful_unit_test_template = Template(
            'Unit test toegevoegd\n',
            '--------------------\n',
            'We hebben deze unit test toegevoegd.',
            'De functie die het externe softwarebedrijf had geschreven bleek inderdaad niet de specificatie te volgen.',
            'Zij hebben de functie verbeterd en nu zou het beter moeten zijn.',
        )
        self.incorrect_unit_test_template = Template(
            'Unit test NIET toegevoegd\n',
            '-------------------------\n',
            'We hebben je unit test naast de specificatie gelegd.',
            'Je unit test blijkt niet correct te zijn.',
            'We hebben de unit test dus niet toegevoegd aan onze code.',
        )
        self.hand_in_unit_tests_template = Template(
            'Unit testen inleveren\n',
            '---------------------\n',
            'Bedankt!',
            'We hebben de laatste versie van de functie in productie gebracht.',
        )
        self.bug_found_template = Template(
            'Foutmelding van een klant\n',
            '-------------------------\n',
            'Een klant heeft een fout in de functie gemeld.',
            'De functie gaf namelijk het volgende foutieve resultaat.',
            '{test_result}',
            'Jouw bijdrage in de kosten om dat te herstellen is {penalty_bug}.',
        )
        self.no_bug_found_template = Template(
            'Geen foutmeldingen meer\n',
            '-----------------------\n',
            'Klanten melden geen fouten meer in de functie.',
            'Bedankt daarvoor!',
        )
        self.end_negative_template = Template(
            'Einde\n',
            '-----\n',
            'Er zitten nog fouten in de functie,',
            'dus we betalen je niets uit.',
        )
        self.end_positive_template = Template(
            'Einde\n',
            '-----\n',
            'Gefeliciteerd! De functie is dankzij jouw unit testen helemaal foutloos.',
            'We betalen je dan ook met plezier {score} uit.',
        )
        self.total_negative_template = Template(
            'Totaal\n',
            '------\n',
            'In totaal heb je {score} verlies geleden.',
            'Jammer!',
            'We hopen dat het de volgende keer beter gaat.',
            'Bedankt voor het spelen!',
        )
        self.total_positive_template = Template(
            'Totaal\n',
            '------\n',
            'In totaal heb je {score} verdiend.',
            'Super!',
            'We denken dat je het de volgende keer nog beter gaat doen.',
            'Bedankt voor het spelen!',
        )
        self.invalid_choice_template = Template(
            'Ongeldige keuze\n',
            '---------------\n',
            'Je hebt ongeldige keuze "{choice}" ingevoerd.',
        )

    def format_score(self, score):
        return f'{"" if score >= 0.0 else "-"}€{round(abs(score) * 10000)}'
    

File: game_dutch.js
"use strict";
class Dutch extends Game {
    constructor() {
        super();
    }
    language() {
        return 'Nederlands';
    }
    choiceLabel() {
        return 'Keuze';
    }
    buttonText() {
        return 'Ga!';
    }
    menuMessage(penaltyHint, penaltyBug, penaltyEnd, form) {
        return new Section([
            new Paragraph('Ik wil…'),
            new Paragraph('[1] …het contract zien'),
            new Paragraph('[2] …de specificatie zien'),
            new Paragraph('[3] …een unit test toevoegen'),
            new Paragraph(`[4] …een hint voor een unit test zien (-${this.formatScore(penaltyHint)})`),
            new Paragraph(`[5] …de unit testen inleveren (-${this.formatScore(penaltyBug)} bij foutmelding)`),
            new Paragraph(`[0] …het spel beëindigen (-${this.formatScore(penaltyEnd)} bij foutmelding)`),
            form.toHtml(),
        ]);
    }
    optionSeeContractMessage() {
        return new Section([
            new Paragraph('Ik wil het contract zien.'),
        ]);
    }
    optionSeeProblemDescriptionMessage() {
        return new Section([
            new Paragraph('Ik wil de specificatie zien.'),
        ]);
    }
    optionSeeHintMessage() {
        return new Section([
            new Paragraph('Ik wil een hint voor een unit test zien.'),
        ]);
    }
    optionSubmitMessage() {
        return new Section([
            new Paragraph('Ik wil de unit testen inleveren.'),
        ]);
    }
    optionEndMessage() {
        return new Section([
            new Paragraph('Ik wil het spel beëindigen.'),
        ]);
    }
    unitTestsPanel(unitTests) {
        const list = unitTests.length === 0
            ? [new Paragraph('Je hebt nog geen unit test geschreven.')]
            : unitTests.map(unitTest => unitTest.toHtml());
        return new Section([
            new Header('Unit testen\n'),
            ...list,
        ]);
    }
    currentCandidatePanel(candidate) {
        return new Section([
            new Header('Huidige functie\n'),
            candidate.toHtml(),
        ]);
    }
    scorePanel(score) {
        return new Section([
            new Header('Verdiensten\n'),
            new Paragraph(`${this.formatScore(score)}`),
        ]);
    }
    contractMessage(initialScore, penaltyBug) {
        return new Section([
            new Paragraph('Wij laten een nieuwe functie ontwikkelen door een extern softwarebedrijf. ' +
                'We hebben jou ingehuurd om te zorgen dat die functie ALTIJD het juiste resultaat geeft. ' +
                'Wat goede resultaten zijn staat beschreven in de specificatie. ' +
                'Jouw taak is om voldoende unit testen te schrijven voor die functie, ' +
                'zodat de functie geen foute resultaten meer kan geven. ' +
                `Voor het hele traject krijg je ${this.formatScore(initialScore)}. ` +
                'In het menu staan bij sommige acties kosten vermeld voor jou. ' +
                'Als een gebruiker bijvoorbeeld een fout constateert in een functie die slaagt voor al jouw unit testen, ' +
                `dan betaal jij een boete van ${this.formatScore(penaltyBug)}.`),
        ]);
    }
    addUnitTestFormMessage(form) {
        return new Section([
            new Paragraph('Ik wil een unit test toevoegen.'),
            form.toHtml(),
        ]);
    }
    addUnitTestTextMessage(unitTest) {
        return new Section([
            new Paragraph('Ik wil de volgende unit test toevoegen:'),
            unitTest.toHtml(),
        ]);
    }
    hintUnitTestMessage(unitTest, penaltyHint) {
        return new Section([
            new Paragraph('Een unit test die nu niet slaagt is bijvoorbeeld de volgende.'),
            unitTest.toHtml(),
            new Paragraph(`De kosten voor deze hint zijn ${this.formatScore(penaltyHint)}.`),
        ]);
    }
    bugFoundMessage(testResult, penaltyBug) {
        return new Section([
            new Paragraph('Bedankt! ' +
                'We hebben de laatste versie van de functie in productie gebracht. ' +
                'Een klant heeft een fout in de functie gemeld. ' +
                'De functie gaf namelijk het volgende foutieve resultaat.'),
            testResult.toHtml(),
            new Paragraph(`Jouw bijdrage in de kosten om dat te herstellen is ${this.formatScore(penaltyBug)}.`),
        ]);
    }
    endWithBugMessage() {
        return new Section([
            new Paragraph('Er zitten nog fouten in de functie, ' +
                'dus we betalen je niets uit. ' +
                'Jammer! ' +
                'We hopen dat het de volgende keer beter gaat. ' +
                'Bedankt voor het spelen!'),
        ]);
    }
    endPerfectMessage(score) {
        return new Section([
            new Paragraph('Gefeliciteerd! ' +
                'De functie is dankzij jouw unit testen helemaal foutloos. ' +
                `In totaal heb je de maximale ${this.formatScore(score)} verdiend. ` +
                'Super! ' +
                'Bedankt voor het spelen!'),
        ]);
    }
    endPositiveMessage(score) {
        return new Section([
            new Paragraph('Gefeliciteerd! ' +
                'De functie is dankzij jouw unit testen helemaal foutloos. ' +
                `In totaal heb je ${this.formatScore(score)} verdiend. ` +
                'Goed gedaan! ' +
                'We denken dat je het de volgende keer nog beter gaat doen. ' +
                'Bedankt voor het spelen!'),
        ]);
    }
    endNegativeMessage(score) {
        return new Section([
            new Paragraph('Gefeliciteerd! ' +
                'De functie is dankzij jouw unit testen helemaal foutloos. ' +
                `In totaal heb je ${this.formatScore(score)} verlies geleden. ` +
                'Jammer! ' +
                'We hopen dat het de volgende keer beter gaat. ' +
                'Bedankt voor het spelen!'),
        ]);
    }
    uselessUnitTestMessage() {
        return new Section([
            new Paragraph('We hebben deze unit test toegevoegd. ' +
                'De huidige functie slaagde al voor deze unit test. ' +
                'We denken daarom dat deze unit test niet zo zinvol is.'),
        ]);
    }
    usefulUnitTestMessage() {
        return new Section([
            new Paragraph('Gedaan.'),
        ]);
    }
    incorrectUnitTestMessage() {
        return new Section([
            new Paragraph('We hebben je unit test naast de specificatie gelegd. ' +
                'Je unit test blijkt niet correct te zijn. ' +
                'We hebben de unit test dus niet toegevoegd aan onze code.'),
        ]);
    }
    formatScore(score) {
        return `${score < 0 ? '-' : ''}€${Math.round(Math.abs(score) * 10)}`;
    }
}
